logbook-json/src/main/java/org/zalando/logbook/json/JsonHttpLogFormatter.java

```diff
@@ -1,92 +1,184 @@
-
-    private final ObjectMapper mapper;
+public final class JsonHttpLogFormatter implements HttpLogFormatter {
 
+    private final JsonFactory jsonFactory;
+    
     public JsonHttpLogFormatter() {
         this(new ObjectMapper());
     }
+    
+    public JsonHttpLogFormatter(JsonFactory jsonFactory) {
+        this.jsonFactory = jsonFactory;
+    }
 
-    public JsonHttpLogFormatter(final ObjectMapper mapper) {
-        this.mapper = mapper;
+    public JsonHttpLogFormatter(ObjectMapper objectMapper) {
+        this.jsonFactory = objectMapper.getFactory();
     }
 
     @Override
-    public Optional<Object> prepareBody(final HttpMessage message) throws IOException {
-        final String contentType = message.getContentType();
-        final String body = message.getBodyAsString();
-        if(body.isEmpty()) {
-            return Optional.empty();
+    public String format(Precorrelation precorrelation, HttpRequest request) throws IOException {
+        final String correlationId = precorrelation.getId();
+        
+        String body = request.getBodyAsString();
+
+        StringBuilderWriter writer = new StringBuilderWriter(body.length() + 2048);
+        
+        try (JsonGenerator generator = jsonFactory.createGenerator(writer)) { 
+            generator.writeStartObject();
+            
+            generator.writeFieldName("origin");
+            if(request.getOrigin() == Origin.LOCAL) {
+                generator.writeRawValue("\"local\"");
+            } else {
+                generator.writeRawValue("\"remote\"");
+            }
+            generator.writeFieldName("type");
+            generator.writeRawValue("\"request\"");
+            generator.writeStringField("correlation", correlationId);
+            generator.writeStringField("protocol", request.getProtocolVersion());
+            generator.writeStringField("remote", request.getRemote());
+            
+            generator.writeStringField("method", request.getMethod());
+
+            generator.writeFieldName("uri");
+            reconstructUri(request, generator, writer);
+            
+            writeHeaders(generator, request);
+    
+            writeBody(request, body, generator);
+    
+            generator.writeEndObject();
         }
-        if (JsonMediaType.JSON.test(contentType)) {
-        	// TODO has this JSON been validated? If not then this might result in invalid log statements
-            return Optional.of(new JsonBody(body));
-        } else {
-            return Optional.of(body);
+        
+        return writer.toString();
+            
+    }
+
+    private void writeBody(HttpMessage message, String body, JsonGenerator builder) throws IOException {
+        if(!body.isEmpty()) {
+            builder.writeFieldName("body");
+
+            final String contentType = message.getContentType();
+
+            if (JsonMediaType.JSON.test(contentType)) {
+                builder.writeRawValue(body);
```

Comment:
This is now way more expressive than in the original version that used `@JsonRawValue`. :+1: